### ISSUE TITLE
Improving memory efficiency

### ISSUE THREAD

**Issue by seanlinsley** [activerecord-hackery/ransack#819]
How can we optimize Ransack's memory usage? I'd like to discuss that here, ideally with multiple PRs created as a result of that discussion.

I discovered with the rack-mini-profiler and memory_profiler gems that Ransack creates a lot of objects in memory while rendering a page, just to be thrown away. I'd started looking into ActiveAdmin memory usage, but discovered that Ransack creates twice the number of objects as ActiveAdmin does.

https://github.com/activeadmin/activeadmin/issues/5113

You can download the [full report](https://github.com/activerecord-hackery/ransack/files/1185151/ransack-memory.txt), but here are the highlights:
```
Total allocated: 1401988 bytes (17704 objects)
Total retained:  1112 bytes (18 objects)

allocated memory by location
-----------------------------------
    462768  ransack-8ca8c95c6543/lib/ransack/predicate.rb:28
    215840  ransack-8ca8c95c6543/lib/ransack/adapters/active_record/base.rb:36
    129708  ransack-8ca8c95c6543/lib/ransack/predicate.rb:22
    120080  ransack-8ca8c95c6543/lib/ransack/predicate.rb:13
    113760  ransack-8ca8c95c6543/lib/ransack/predicate.rb:9
    107920  ransack-8ca8c95c6543/lib/ransack/adapters/active_record/base.rb:37
     98448  ransack-8ca8c95c6543/lib/ransack/adapters/active_record/base.rb:48
     30552  ransack-8ca8c95c6543/lib/ransack/adapters/active_record/ransack/context.rb:41


allocated objects by location
-----------------------------------
     11190  ransack-8ca8c95c6543/lib/ransack/predicate.rb:28
      2058  ransack-8ca8c95c6543/lib/ransack/adapters/active_record/base.rb:48
       730  ransack-8ca8c95c6543/lib/ransack/predicate.rb:22
       568  ransack-8ca8c95c6543/lib/ransack/adapters/active_record/base.rb:36
       330  ransack-8ca8c95c6543/lib/ransack/context.rb:138
       316  ransack-8ca8c95c6543/lib/ransack/predicate.rb:13
       284  ransack-8ca8c95c6543/lib/ransack/adapters/active_record/base.rb:37
```

This is the query string I used on an ActiveAdmin index page of my app:
```
?pp=profile-memory&memory_profiler_allow_files=lib/ransack
```
Refer to the [docs](https://github.com/MiniProfiler/rack-mini-profiler#memory-profiling) for more options.

This is an important issue to fix because it causes unnecessary GC pressure on the thousands of sites that use Ransack, with or without ActiveAdmin. And if I understand it correctly, retained objects are potential memory leaks.

Memoization is by far the easiest fix when it's possible. `String#freeze` and the [frozen string pragma](https://wyeworks.com/blog/2015/12/1/immutable-strings-in-ruby-2-dot-3) can also help, but they might cause bugs.

**Comment by seanlinsley** [comment]
These are the Ransack customizations this app has:
```ruby
Ransack.configure do |config|
  # https://github.com/activerecord-hackery/ransack/issues/321
  {
    array_contained_within:           :contained_within,
    array_contained_within_or_equals: :contained_within_or_equals,
    array_contains:                   :contains,
    array_contains_or_equals:         :contains_or_equals,
    array_overlap:                    :overlap
  }.each do |rp, ap|
    config.add_predicate rp, arel_predicate: ap, wants_array: true
  end

  # https://github.com/activerecord-hackery/ransack/issues/580
  config.sanitize_custom_scope_booleans = false
end
```

**Comment by seanlinsley** [comment]
[`detect_from_string`](https://github.com/activerecord-hackery/ransack/blob/8ca8c95c6543/lib/ransack/predicate.rb#L28) represents 63% of allocations but only 33% of memory.

[`ransackable_attributes`](https://github.com/activerecord-hackery/ransack/blob/8ca8c95c6543c5da9b98a4e2747a0a4023d60bfc/lib/ransack/adapters/active_record/base.rb#L36-L37) only represents 5% of allocations and yet takes up 23% of memory.

**Comment by corsonknowles** [comment]
What's the status on this? We just upgraded to ActiveAdmin 1.1 and immediately experience 1gb memory usage from its Ransack dependency. We were not having the same issue with ActiveAdmin's use of Metasearch in older builds.

**Comment by seanlinsley** [comment]
1 GB? That's alarming if true. Could you profile the memory usage in your app as documented here?

I'm currently using #820 in production to see if it causes any bugs.

**Comment by corsonknowles** [comment]
Thanks Sean! Looks like we wrote some custom ransackers to replace what Metasearch was doing and they ended up loading all of the data we were interested in searching into memory (in this case User Support conversation histories), which was a 1gb file for each load. We've removed them and that solves the issue for now. I'm going to keep an eye on the other two ransackers we wrote to replace Metasearch, they seem to have only minor memory issues (certainly minor by comparison!)
